### ISSUE TITLE
Possible contradiction in rules for naming source distribution files containing `-_.`

### ISSUE THREAD

**Issue by priolacci** [pypa/packaging.python.org#1750]
### Issue Description

[The specification page for source distributions](https://packaging.python.org/en/latest/specifications/source-distribution-format/) states the following:

> The file name of a sdist was standardised in PEP 625. The file name must be in the form {name}-{version}.tar.gz, where {name} is normalised according to the same rules as for binary distributions (see Binary distribution format), and {version} is the canonicalized form of the project version (see Version specifiers).
> 
> The name and version components of the filename MUST match the values stored in the metadata contained in the file.

The [rules for binary distributions](https://packaging.python.org/en/latest/specifications/binary-distribution-format/#escaping-and-unicode) state that `-_.` should be replaced by `_` in distribution names.

But the [naming convention for core metadatas](https://packaging.python.org/en/latest/specifications/core-metadata/#name) requires the package maintainer to follow [this normalization](https://packaging.python.org/en/latest/specifications/name-normalization/#name-normalization), which states `-_.` should be replaced by `-`.

Hence, for a package named for instance "my-package" v0.1, the metadata specification would require the "Name" attribute to be `my-package`, while the specification for source distribution files requires the name of the file to be `my_package-0.1.tar.gz`, making it impossible to follow this part of the source distribution filename specification: _"The name and version components of the filename MUST match the values stored in the metadata contained in the file."_

### Code of Conduct

- [X] I am aware that participants in this repository must follow the PSF Code of Conduct.

**Comment by webknjaz** [comment]
I suppose "match" != "be equal exactly"? It might mean "correspond" in this context.

**Comment by abravalheri** [comment]
Yes I agree with @zahlman, there is a difference between the package naming as provided by the (human) user[^1] and the file naming as intended to be consumed by (automated) tools.

The wheel and sdist file names need to be optimised so that the installers can quickly infer the version. That is why the package name need to be "normalised" by replacing `.` and `-` with `_` when creating the archive. This way the heuristic for inferring the version is simply a variation of `filename.split('-')[1]`.

[^1]: and it may contain `.-` characters

**Comment by abravalheri** [comment]
> Hence, for a package named for instance "my-package" v0.1, the metadata specification would require the "Name" attribute in the metadata to be my-package

This normalisation when writing core metadata files is not necessary. There was a discussion at the time of the PEPs proposal and adoption about this, the conclusion is that the metadata should preserve the intent of the user (how they want to call the package), but tools should normalising it when comparing or processing the names in automated processes.

There are many related topics in the Discourse, but this is a short one summarising the discussions https://discuss.python.org/t/revisiting-distribution-name-normalization/12348/15.
